content/6_Advanced/Max_Flow.mdx

```diff
@@ -112,6 +112,140 @@ int main() {
 
 <Problems problems="match" />
 
+It is more than recommended to solve the first problem - Download Speed - in the section before jumping to this one.
+
+### Solution 1
+
+A bipartite graph doesn't seem like to have anything in common with a flow network, but we can shift our point of view just by
+adding the source connected to the nodes in set $U$ and the sink connected to the nodes in set $V$. And that's all.
+Now we have a flow network where every capacity is equal to 1.
+
+The reason we transformed our bipartite graph into a network flow is that the maximum network flow is equal
+to the maximum matching. We can now apply Edmonds-Karp or Ford-Fulkerson algorithm.
+
+**Time Complexity: $\mathcal{O}(VE^2)$**
+
+### Solution 2
+
+We can achieve a better complexity than $\mathcal{O}(VE^2)$, but first let's define some properties of matching algorithms.
+
+Let's say the set $M$ contains all the edges that the maximum matching consists of.
+
+We define an **alternating path** as a path whose edges are in the matching, $M$, and not in the matching, in an alternating fashion. An alternating path stars with an unmatched node and ends once it cannot append another edge while maintaining an alternating sequence.
+
+An **augmenting path** is built upon the alternating path and unmatched nodes at both ends - the nodes are not included in $M$.
+
+Tha maximum matching $M$ can be further improved if and only if an augmenting path is found in $M$, otherwise $M$ is the maximum matching. It may seem difficult to understand, but the main idea is: ``The maximum matching can be further improved as long as the alternating sequence can be extended ``. For a better understanding, you can imagine the shoelaces as an alternating path in a bipartite graph - or the sneakers.
+
+The algorithm described above is called the [Hopcroft-Karp algorithm](https://en.wikipedia.org/wiki/Hopcroft%E2%80%93Karp_algorithm).
+
+**Time Complexity: $\mathcal{O}(E\sqrt{V})$**
+
+Here is an animation of how the code below works:
+
+<video width="960" height="720" controls>
+	<source src="/animations/Hopcroft-Karp.mp4" type="video/mp4"/>
+</video>
+
+<LanguageSection>
+<CPPSection>
+```cpp
+#include <bits/stdc++.h>
+
+using namespace std;
+
+const int INF = 1e9;
```

Comment:
int_max maybe?

Reply:
It's safer not tu put INT_MAX, could overflow